### ISSUE TITLE
Secret key for encode

### ISSUE THREAD

**Issue by devgowa** [devopsfaith/krakend-jose#19]
How this extra plugin krakend work with **secret key** for encode payload of response from another services? I need to know what krakend-jose use **secret key** for validate token from another service with another programing language. Thanks before.

**Comment by kpacha** [comment]
hi, @devgowa and welcome to the KrakenD community!

I'm not sure if I got your question right... this module does not encrypt/decrypt the JWT, it just signs and verifies the signature.

Signing JWT can be done with symmetric and asymmetric keys.
- Symmetric: same key for signing and verifying (HSA256, etc)
- Asymmetric: sign with the private key and verify with the public key (RSA256, etc)

Those keys are provided to the gateway using a remote service exposing JWK

For a complete overview, please check the documentation: https://www.krakend.io/docs/authorization/jwt-overview/

cheers!

**Comment by devgowa** [comment]
Yes, I completely success with this https://www.krakend.io/docs/authorization/jwt-signing/ method for signing a JWT, but I want use that JWT into another programing language. For example, I wanna to verify that JWT token to PHP programing language with this package https://github.com/firebase/php-jwt, can I do that?

Thank's before.

**Comment by kpacha** [comment]
It's been a while since the last time I wrote some PHP code but, at a first glance, it looks like the lib you're linking already supports 'HS256' and 'RS256'.

**Comment by devgowa** [comment]
Yeah, I know it should be 'easy' but to verify JWT token which signed by **krakend-jose** I need secret key, I read your code in here https://github.com/devopsfaith/krakend-jose/blob/master/jws.go I can't find any secret key. 

Thank's.

**Comment by kpacha** [comment]
**TL;DR**: there is no secret in this lib. you should provide the secrets for signing (either the HSA key or the private RSA key) and then you should provide the same HSA key or the public RSA key to your token verifiers (depending of the algorithm you select)

You said you already succeeded generating signed JWT with the KrakenD, so you are already using some algorithm and key. If you are using HSA, then you already have the secret. If you selected RSA, you need the public key associated to the private one used at the signing stage.

The keys included at the JWK are base64 encoded. You should check how to inject them to your verifier.

**Comment by devgowa** [comment]
Thank's for the explaination, I will try to find out. 
How about manipulate response of **krakend-jose**?
I need to showing out JSON format with any status code. 
Can I do that?

**Comment by alombarte** [comment]
Hi @devgowa, I think it's going to be more efficient to ask your questions in Slack, as this is not a issue on the JOSE library.

You can join https://invite.slack.golangbridge.org/ and ask in the `#krakend` channel.

Thanks
